src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -0,0 +1,309 @@
+        }
+        /// <summary>
+        /// Update the Activity to set <see cref="Duration"/>
+        /// as a difference between <see cref="StartTimeUtc"/>
+        /// and given stop timestamp
+        /// </summary>
+        /// <param name="endTimeUtc">Activity stop time in UTC (Greenwitch Mean Time)</param>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity SetEndTime(DateTime endTimeUtc)
+        {
+            Duration = endTimeUtc - StartTimeUtc;
+            return this;
+        }
+
+        /// <summary>
+        /// If the Activity has ended (Stop was called) then this is the delta
+        /// between start and end.   If the activity is not ended then this is 
+        /// TimeSpan.Zero.  
+        /// </summary>
+        public TimeSpan Duration { get; private set; }
+
+        /// <summary>
+        /// Starts activity: sets <see cref="Parent"/> to hold <see cref="Current"/> and sets Current to this Activity.
+        /// If <see cref="StartTimeUtc"/> was not set previously, sets it to DateTime.UtcNow.
+        /// Use DiagnosticSource.Start to start activity and write start event.
+        /// </summary>
+        /// <returns>Started activity for convinient chaining</returns>
+        /// <seealso cref="DiagnosticSource.StartActivity(Activity, object)"/>
+        /// <seealso cref="SetStartTime(DateTime)"/>
+        public Activity Start()
+        {
+            if (Id != null)
+                throw new InvalidOperationException("Trying to start an Activity that was already started");
+
+            if (ParentId == null)
+            {
+                var parent = Current;
+                if (parent != null)
+                {
+                    ParentId = parent.Id;
+                    Parent = parent;
+                }
+            }
+
+            if (StartTimeUtc == default(DateTime))
+                StartTimeUtc = DateTime.UtcNow;
+
+            Id = GenerateId();
+
+            Current = this;
+            return this;
+        }
+
+        /// <summary>
+        /// Stops activity: sets Current to Parent.
+        /// If end time was not set previously, sets <see cref="Duration"/> as a difference between DateTime.UtcNow and <see cref="StartTimeUtc"/>
+        /// Use DiagnosticSource.Stop to stop activity and write stop event.
+        /// </summary>
+        /// <seealso cref="DiagnosticSource.StopActivity(Activity, object)"/>
+        /// <seealso cref="SetEndTime(DateTime)"/>
+        public void Stop()
+        {
+            if (Id == null)
+                throw new InvalidOperationException("Trying to stop an Activity that was not started");
+
+            if (!isFinished)
+            {
+                isFinished = true;
+
+                if (Duration == TimeSpan.Zero)
+                    SetEndTime(DateTime.UtcNow);
+
+                Current = Parent;
+            }
+        }
+
+        #region private 
+
+        private string GenerateId()
+        {
```

Comment:
Weird algorithm. 

As far as I've seen, most tracers use e.g. 128bit random numbers to be able to easily compute sampling rates. (1% of all requests means that all numbers below a threshold will be tracked)

Also, why have different logic for DEBUG/RELEASE?